### PR TITLE
Stop considering Rectangle(0,0,0,0) null - while Rectangle(1,1,1,1) is not null

### PR DESCRIPTION
**Pull request by strk**
Superceeds GH-53323 

Aims at closing GH-45563

### CODE REVIEW

File: src/core/geometry/qgsrectangle.cpp
Comment (strk) [reply src/core/geometry/qgsrectangle.cpp]: Thanks, I've moved the code up. As for ::asWktPolygon I'm trying to keep the scope of this PR low and not change too many things. The focus of this PR is fixing a bug which makes `Rectangle(0,0,0,0)` considered in a different way from `Rectangle(1,1,1,1)`. It doesn't make sense to me that one is considered "null" while the other is not.

Whether it does make sense to have "null" and "empty" represent different things I'm not sure but it's not in the scope of this PR to change.

File: src/core/geometry/qgsrectangle.h
Comment (strk) [reply src/core/geometry/qgsrectangle.h:615]: I believe clients are using setMinimal() to effectively mean setNull() (which is missing as a method) and the values we set in the default constructor should be matching those values. Do you have any idea why the CI is failing ? That's the current goal (making CI happy)

Comment (strk) [reply src/core/geometry/qgsrectangle.h:615]: See what I mean here:
https://github.com/qgis/QGIS/actions/runs/6378688541/job/17315227587?pr=54646#step:13:3625

I don't seem to be able to run the test locally:
```
[strk@c19:/usr/local/src/qgis/qgis/src/rectangle-null/build(rectangle-null)] ctest  -R  PyQgsLayoutItemMapItemClipPathSettings --output-on-failure
Test project /usr/local/src/qgis/qgis/src/rectangle-null/build
    Start   1: logGitStatus
1/3 Test   #1: logGitStatus .............................   Passed    0.02 sec
    Start 572: PyQgsLayoutItemMapItemClipPathSettings
2/3 Test #572: PyQgsLayoutItemMapItemClipPathSettings ...***Failed    0.23 sec
export =
export PYTHONPATH=/usr/local/src/qgis/qgis/src/rectangle-null/build/output/python:/usr/local/src/qgis/qgis/src/rectangle-null/build/output/python/plugins:/usr/local/src/qgis/qgis/src/rectangle-null/tests/src/python::/opt/gdal-3.6/lib/python3/dist-packages/:/opt/gdal-3.6/lib/python3/dist-packages/
-- Running  /usr/bin/python3.10 /usr/local/src/qgis/qgis/src/rectangle-null/tests/src/python/test_qgslayoutmapitemclippingsettings.py
terminate called after throwing an instance of 'QgsSettingsException'
CMake Error at /usr/local/src/qgis/qgis/src/rectangle-null/build/tests/src/python/PyQgsLayoutItemMapItemClipPathSettings.cmake:23 (message):
  Test failed: Subprocess aborted



    Start   2: checkGitStatus
3/3 Test   #2: checkGitStatus ...........................   Passed    0.02 sec

67% tests passed, 1 tests failed out of 3

Total Test time (real) =   0.28 sec

The following tests FAILED:
        572 - PyQgsLayoutItemMapItemClipPathSettings (Failed)
Errors while running CTest
```

Comment (strk) [reply src/core/geometry/qgsrectangle.h:615]: Ok I've resolved my local problem with a clean build, I can reproduce locally the PyQgsLayoutItemMapItemClipPathSettings failure, now to figure out why we end up with a NaN rectangle when a valid one is expected

Comment (strk) [reply src/core/geometry/qgsrectangle.h:615]: The test is weird in that it uses this construct:

>        map.zoomToExtent(QgsRectangle(100, 200, 50, 40))

The QgsRectangle constructor is expecting xmin as the first argument and xmax as the third argument, here we're passing a MIN=100 and MAX=50 which is swapped, is that intentional ? @nyalldawson the code seem to be yours, is that intentional to swap min/max for a rectangle there ? There's no comment helping to figure that out

Comment (strk) [reply src/core/geometry/qgsrectangle.h:615]: Assuming the swapped min/max is not a problem (shouldn't be), the ractangle passed to attemptSetSceneRect (orange), zoomToExtent (purple)  and expected result from toMapClippingRegion are in this picture:

image

I cannot really understand how we get from the input rectangles to the output one but surely a NaN rectangle (that we get) is not good.

Comment (strk) [reply src/core/geometry/qgsrectangle.h:615]: Found the culprit being QgsLayoutItemMap::zoomToExtent computation of aspect ratio guarded by `! rect.isNull()` and not by `! rect.isEmpty()` , which resulted in a division by zero -> NaN. This is fixed by commit 093bb39ad008825e6102e2b434045673a50b35ae in this branch

**Comment by strk** [comment]
Next failure is TestQgsLabelingEngine::testRegisterFeatureUnprojectible in tests/src/core/testqgslabelingengine.cpp which is reported as being a test for bug https://github.com/qgis/QGIS/issues/23431

It looks like that test was written by myself in 4ed096b94dff3101786e67a49567544f77c759d8 so I can't blame anyone else for missing comments (oops :)

**Comment by strk** [comment]
Difference spotted is again an Empty vs. Null difference, in 

tests/src/core/testqgslabelingengine.cpp:947 : (testRegisterFeatureUnprojectible) [1ms] vl2->extent(): Null

vl2->extent() prints as Empty in master branch, as Null in the branch associated with this PR

Honestly I don't know what the correct answer should be, as vl2 is a QgsVectorLayer started empty. An EMPTY Rectangle is defined as being a rectangle with no area (a point, at this moment, which I'd also extend to a line). But an empty layer cannot have a point information in it. I start wondering if we really need the distinction between EMPTY and NULL, for a rectangle.

File: src/core/geometry/qgsrectangle.cpp
Comment (strk) [reply src/core/geometry/qgsrectangle.cpp]: > By the way, you should also modify the `QgsRectangle::asWktPolygon` function as it should return `POLYGON EMPTY` when `isNull` is TRUE (cf. https://libgeos.org/specifications/wkt/).

I've done this in GH-54831 - could you please review/approve ? Thanks

File: tests/src/python/providertestbase.py
Comment (strk) [inline tests/src/python/providertestbase.py:454]: But I don't think this should be the case

Comment (nyalldawson) [reply tests/src/python/providertestbase.py:454]: I think it should be null for an empty layer too, can you change that please?

Comment (strk) [reply tests/src/python/providertestbase.py]: Rebased to master and removed this commit, I'll have to find out why the tests end up getting an empty but not null extent

**Comment by strk** [comment]
Next issue is with a WMS test of QGIS server that expects a GetFeatureInfoResponse having this bounding box:
```
 <BoundingBox maxy="0" maxx="0" miny="0" CRS="EPSG:3857" minx="0"/>
```

We're now outputting the -inf,+inf bounding box instead.

The question is: what is expected as a BoundingBox for a response with an empty bounding box ?

The file in question is https://github.com/qgis/QGIS/blob/master/tests/testdata/qgis_server/get_postgres_types_json_list.txt

@elpaso should we just omit the BoundingBox tag when the box is null ? What is the reference spec ?

**Comment by strk** [comment]
I've filed PR https://github.com/qgis/QGIS/pull/54858 to address the null BoundingBox in WMS output

File: src/core/geometry/qgsrectangle.cpp
Comment (lnicola) [inline src/core/geometry/qgsrectangle.cpp:148]: Shouldn't we still handle the `isEmpty` case here?

Comment (strk) [reply src/core/geometry/qgsrectangle.cpp:148]: I don't think so, or at least it's not specifically useful as the definition of an "empty" rectangle is just that the rectangle has no area (collapsed to a point or vertical or horizontal line)

Comment (strk) [reply src/core/geometry/qgsrectangle.cpp:148]: Of course this is being a problem anyway with null... eh:
https://github.com/qgis/QGIS/actions/runs/6429563312/job/17464475280?pr=54646#step:13:540

Comment (lnicola) [reply src/core/geometry/qgsrectangle.cpp:148]: But if QGIS has two different concepts for "null" and "empty" rectangles, shouldn't it be visible in the `toString` output? :shrug:

File: src/core/geometry/qgsrectangle.h
Comment (lnicola) [inline src/core/geometry/qgsrectangle.h:525]: This seems broken for non-null rectangles (`false == false`).

File: src/core/geometry/qgsrectangle.cpp
Comment (strk) [reply src/core/geometry/qgsrectangle.cpp:148]: Empty is not strictly a concept, more of a characteristic. It's like saying area is zero. The coordinates give the whole detail, there's no need for special handling because there isn't the problem of having to represent NaN numbers or max_double or similar when a supposedly common "null" rectangle is encountered.

**Comment by nyalldawson** [comment]
@strk

Can you open PRS for https://github.com/qgis/QGIS/pull/54646/commits/0dc22c1569dff21382ce100b228e7639c93577b2, 
https://github.com/qgis/QGIS/pull/54646/commits/d5d9e5c557d2dce972383513a9352f0cd39f988d,
https://github.com/qgis/QGIS/pull/54646/commits/6372baec5be8efc72c2db77e463cd4819fc52781
And
https://github.com/qgis/QGIS/pull/54646/commits/c74837ade617fc23b2d5275b99127a0bc6d70cb5

?

Those can all be merged now

**Comment by strk** [comment]
> @strk
> 
> Can you open PRS for [0dc22c1](https://github.com/qgis/QGIS/commit/0dc22c1569dff21382ce100b228e7639c93577b2), [d5d9e5c](https://github.com/qgis/QGIS/commit/d5d9e5c557d2dce972383513a9352f0cd39f988d), [6372bae](https://github.com/qgis/QGIS/commit/6372baec5be8efc72c2db77e463cd4819fc52781) And [c74837a](https://github.com/qgis/QGIS/commit/c74837ade617fc23b2d5275b99127a0bc6d70cb5)
> 
> ?
> 
> Those can all be merged now

Done in GH-54872 - I will rebase this one once that one is merged

File: tests/src/python/providertestbase.py
Comment (nyalldawson) [reply tests/src/python/providertestbase.py]: Yeah, I'd say the provider is wrong there

Comment (strk) [reply tests/src/python/providertestbase.py]: I get our friend 0,0,0,0 empty rect instead of the null one:

> AssertionError: <QgsRectangle: 0 0, 0 0> != <QgsRectangle: 0 1, 2 3>

Still trying to figure out why

Comment (strk) [reply tests/src/python/providertestbase.py]: Ok figured and fixed, confirming by merging [PR 54917 ](https://github.com/qgis/QGIS/pull/54917) into this branch. If you can please approve that one I'll merge and rebase this after dealing with the same issue with the AFS provider.

**Comment by strk** [comment]
I don't understand the "testExtent" test from tests/src/python/providertestbase.py:
```
    def testExtent(self):
        reference = QgsGeometry.fromRect(
            QgsRectangle(-71.123, 66.33, -65.32, 78.3))
        provider_extent = self.source.extent()
        self.assertAlmostEqual(provider_extent.xMinimum(), -71.123, 3)
```

How is assigning the "reference" local variable expected to affects self.source.extent() ?

**Comment by strk** [comment]
The reference extent test is now ok, next failure is with the ArcGIS rest provider. I pushed a fix for that both in this branch and in a separate branch to make it easier to review the single fix: https://github.com/qgis/QGIS/pull/54944

**Comment by strk** [comment]
Next failure:
```
FAIL!  : TestQgsLabelingEngine::testRegisterFeatureUnprojectible() Compared values are not the same
   Actual   (provider->mLabels.size()): 1
   Expected (0)                       : 0
   Loc: [/root/QGIS/tests/src/core/testqgslabelingengine.cpp(954)]
```

Single Unprojectible  feature is expected to result in no labels, we get 1 label instead.
This also happens in GH-54934 which shares with this PR the change in default-constructed QgsRectangle to not be 0,0,0,0 anymore.

**Comment by strk** [comment]
@nyalldawson let me know if you prefer the last  3 commits to be in their own PR, but note that all these bugs are not exposed until we switch the default QgsRectangle constructor to result in a null (rather than 0,0,0,0) rectangle.

**Comment by strk** [comment]
Next failure here:
```
ctest -R PyQgsVirtualLayerProvider --output-on-failure
...
FAIL: testExtentSubsetString (__main__.TestQgsVirtualLayerProvider)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/src/qgis/qgis/src/rectangle-null/tests/src/python/providertestbase.py", line 462, in testExtentSubsetString
    self.assertEqual(provider_extent, QgsRectangle())
AssertionError: <QgsRectangle: 0 0, 0 0> != <QgsRectangle: 179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368 179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368, -179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368 -179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368>
```

**Comment by strk** [comment]
Nyall since this is now working, and has the default QgsRectangle constructor construct a null rectangle, should I re-move the ::createNull() in this PR ? I'd lean toward yes :)

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Woohoo! Nice work @strk!

**Comment by agiudiceandrea** [comment]
Hi @strk, it seems to me this PR may have generated an issue in the "TIN interpolation" and  "IDW interpolation" QGIS processing algorithms. Please see https://github.com/qgis/QGIS/issues/55138.

**Pull request merged**
